Implement `take` and `slice` in VectorIterator

szeiger opened the pull request:
...and fix the `drop` implementation which didn’t ensure that the
remaining size is reported as 0 when all remaining elements are dropped.

Fixes https://github.com/scala/bug/issues/9954

szeiger committed:
Implement `take` and `slice` in VectorIterator

...and fix the `drop` implementation which didn’t ensure that the
remaining size is reported as 0 when all remaining elements are dropped.

Fixes https://github.com/scala/bug/issues/9954

diesalbla commented on test/junit/scala/collection/immutable/VectorTest.scala:
```diff
@@ -181,4 +181,41 @@ class VectorTest {
     val x: Vector[Char] = Vector[Char]().tapEach(f)
   }
+
+  @Test
```
I think that, using property tests of Scalacheck, these tests can be written as:  
 ```Scala
prop { (v: Vector[Int], i: Int, j: Int) => 
    v.take(i).toArray.take(j) === v.take(i).iterator.take(j).toArray
}
```
That may be a wider set of tests than these ones.

szeiger replied:
The existing tests were JUnit so that's what I used. I was hoping for scalacheck tests, too.

diesalbla commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -659,6 +659,16 @@ class VectorIterator[+A](_startIndex: Int, endIndex: Int)
 
+  override def take(n: Int): Iterator[A] = {
+    val dropR = remainingElementCount - n
+    if(dropR > 0) {
```
IMHO the following condition better expresses that this does no change unless "we take less than we have":  
```suggestion
    if(n < remainingElementCount) {
```
On the other hand, I am not sure if this would fully work with a negative  `n`. Further `take` operations would simply not change the state, still, do we want to leave the iterator on a dirty state, with `_endIndex_ < _startIndex`?

szeiger replied:
Right, this needs to be guarded against a negative n.

I originally had `n < remainingElementCount` but the form in which it is needed afterwards is `remainingElementCount - n`. I think it also reads nicely: "If there is something to drop from the right, drop it".

`endIndex < startIndex` should be fine. I didn't investigate if `Vector` itself can deal with it, which it would if constructed via `VectorIterator#remainingVector` (which appears to be unused -- maybe it's used by the parallel collections) but the code in `VectorIterator` can. `remainingVector` should either be removed or changed to return `Vector.empty` in this case.

szeiger commented:
Updated:
- rebased
- added check for `n <= 0` to `take`
- return `Vector.empty` in `remainingVector` instead of building a new, empty `Vector` with an illegal `endIndex`
